opacus/autograd_grad_sample.py

```diff
@@ -238,4 +238,26 @@ def _compute_grad_sample(
     compute_layer_grad_sample = _supported_layers_grad_samplers.get(
         get_layer_type(layer)
     )
-    compute_layer_grad_sample(layer, A, B)
+
+    if 'max_batch_len' in signature(compute_layer_grad_sample).parameters: # For compatibility with PackedSequences
+        max_batch_len = _get_batch_size(layer, A, batch_dim)
+        compute_layer_grad_sample(layer, A, B, max_batch_len=max_batch_len)
+    else:
+        compute_layer_grad_sample(layer, A, B)        
+
+
+def _get_batch_size(layer, A, batch_dim):
+    if batch_dim != 0:
+        return 0
+
+    if hasattr(layer, 'max_batch_len'):
+        return layer.max_batch_len
+
+    max_batch_len = 0
+    for out in layer.activations:
+        if out.shape[0] > max_batch_len:
+            max_batch_len = out.shape[0]
+
+    max_batch_len = max(max_batch_len, A.shape[batch_dim])
+    layer.max_batch_len = max_batch_len
```

Comment:
I am not sure how this can be done for one batch only. For now, I have removed caching. If in the future this part gets object oriented, the batch-level caching should be easily done. Its quadratic in the number of sequence lengths. The overhead should not be noticeable when dealing with large batch sizes and/or large input dimensions (since these parts will dominate when updating ```param.grad_sample```). For example, I just benchmarked the ```_get_batch_size``` function on an activations list of 1 million entries and found that the function took only about 160ms for one call. We rarely see sequence lengths of a million, if at all!

Reply:
I was thinking, that caching could work if we reset the cache on `zero_grad()`?
But I agree, that the overhead shouldn't be too bad and we don't have to solve it now